metadata-io/src/main/java/com/linkedin/metadata/search/utils/ESUtils.java

```diff
@@ -267,46 +267,71 @@ public static void buildSortOrder(
       @Nonnull SearchSourceBuilder searchSourceBuilder,
       @Nullable SortCriterion sortCriterion,
       List<EntitySpec> entitySpecs) {
-    if (sortCriterion == null) {
+    buildSortOrder(
+        searchSourceBuilder,
+        sortCriterion == null ? List.of() : List.of(sortCriterion),
+        entitySpecs,
+        true);
+  }
+
+  /**
+   * Allow disabling default sort, used when you know uniqueness is present without urn field. For
+   * example, edge indices where the unique constraint is determined by multiple fields (src urn,
+   * dst urn, relation type).
+   *
+   * @param enableDefaultSort enable/disable default sorting logic
+   */
+  public static void buildSortOrder(
+      @Nonnull SearchSourceBuilder searchSourceBuilder,
+      @Nonnull List<SortCriterion> sortCriterion,
+      List<EntitySpec> entitySpecs,
+      boolean enableDefaultSort) {
+    if (sortCriterion.isEmpty() && enableDefaultSort) {
       searchSourceBuilder.sort(new ScoreSortBuilder().order(SortOrder.DESC));
     } else {
-      Optional<SearchableAnnotation.FieldType> fieldTypeForDefault = Optional.empty();
-      for (EntitySpec entitySpec : entitySpecs) {
-        List<SearchableFieldSpec> fieldSpecs = entitySpec.getSearchableFieldSpecs();
-        for (SearchableFieldSpec fieldSpec : fieldSpecs) {
-          SearchableAnnotation annotation = fieldSpec.getSearchableAnnotation();
-          if (annotation.getFieldName().equals(sortCriterion.getField())
-              || annotation.getFieldNameAliases().contains(sortCriterion.getField())) {
-            fieldTypeForDefault = Optional.of(fieldSpec.getSearchableAnnotation().getFieldType());
-            break;
-          }
-        }
-        if (fieldTypeForDefault.isPresent()) {
-          break;
-        }
-      }
-      if (fieldTypeForDefault.isEmpty()) {
-        log.warn(
-            "Sort criterion field "
-                + sortCriterion.getField()
-                + " was not found in any entity spec to be searched");
-      }
-      final SortOrder esSortOrder =
-          (sortCriterion.getOrder() == com.linkedin.metadata.query.filter.SortOrder.ASCENDING)
-              ? SortOrder.ASC
-              : SortOrder.DESC;
-      FieldSortBuilder sortBuilder =
-          new FieldSortBuilder(sortCriterion.getField()).order(esSortOrder);
-      if (fieldTypeForDefault.isPresent()) {
-        String esFieldtype = getElasticTypeForFieldType(fieldTypeForDefault.get());
-        if (esFieldtype != null) {
-          sortBuilder.unmappedType(esFieldtype);
-        }
-      }
-      searchSourceBuilder.sort(sortBuilder);
+      sortCriterion.forEach(
+          sortCriteria -> {
+            Optional<SearchableAnnotation.FieldType> fieldTypeForDefault = Optional.empty();
+            for (EntitySpec entitySpec : entitySpecs) {
```

Comment:
nit: Kind of weird to see functional forEach intermixed with traditional forEachs